GPU Parallel Ensemble simulation with random time span

Hi, for my application I would like to simulate a large number of trajectories (1000 to 20000) of the same system with different parameters but also with different time spans for a Monte Carlo simulation. However when attempting this, I get an error from DiffEqGPU.jl
 `ERROR: LoadError: AssertionError: all((p->begin
            #= /home/tangui/.julia/packages/DiffEqGPU/Ibo20/src/DiffEqGPU.jl:278 =#
            p.tspan == (probs[1]).tspan
        end), probs)`
I understand that the function checks if all problems have the same tspan, but why is that ? I guess this has to do with some of the way the problem is broadcasted to the GPU under some fixed array shape. I think I could greatly benefit from switching to computing on GPU, especially for the reduced precision performance so DiffEqGPU seemed to be a good fit. 
Many thanks !
Here is the code to reproduce the error :
```julia
using DiffEqGPU, OrdinaryDiffEq

# global variables to be set from python side through Julia "Main" namespace
N_real = 5000
N_grid = 256
T = 150+273
m87 = 1.44316060e-25
k_B = 1.38064852e-23
window = 2.5e-3
Gamma = 1.0 + 0*im
Omega13 = 1.0 + 0*im
Omega23 = 1.0 + 0*im
gamma21tilde = 1.0 + 0*im
gamma31tilde = 1.0 + 0*im
gamma32tilde = 1.0 + 0*im
waist = 1.0e-3 + 0*im
r0 = 1.0 + 0*im
x0 = ComplexF32[5/8 + 0*im, 3/8+ 0*im, 0+ 0*im, 0+ 0*im, 0+ 0*im, 0+ 0*im, 0+ 0*im, 0+ 0*im]
k = Float32(2*pi/780.241e-9)
v = 40.0f0

function choose_points()::Tuple{Int32, Int32, Int32, Int32}
    edges = Array{Tuple{Int32, Int32}, 1}(undef, 4*N_grid)
    for i in 1:N_grid
        edges[i] = (1, i)
        edges[i+N_grid] = (N_grid, i)
        edges[i+2*N_grid] = (i, 1)
        edges[i+3*N_grid] = (i, N_grid)
    end
    iinit, jinit = edges[rand(1:length(edges), 1)][1]
    ifinal, jfinal = iinit, jinit
    cdtn = (ifinal == iinit) || (jfinal == jinit)
    while cdtn
        ifinal, jfinal = edges[rand(1:length(edges), 1)][1]
        cdtn = (ifinal == iinit) || (jfinal == jinit)
    end
    return (iinit, jinit, ifinal, jfinal)
end

function draw_vz(v::Float32)::Float32
    vz = abs(2*v)
    while abs(vz) > abs(v)
        vz = randn()*sqrt(k_B*T/m87)
    end
    return vz
end

function f!(dy::Array{ComplexF32, 1}, x::Array{ComplexF32, 1},
            p::Array{ComplexF32, 1}, t::Float32)
    v, u0, u1, xinit, yinit = p[1], p[2], p[3], p[4], p[5]
    Gamma, Omega13, Omega23 = p[6], p[7], p[8]
    gamma21tilde, gamma31tilde, gamma32tilde = p[9], p[10], p[11]
    waist, r0 = p[12], p[13]
    r_sq = (xinit+u0*v*t - r0)^2 + (yinit+u1*v*t - r0)^2
    Om23 = Omega23 * exp(-r_sq/(2*waist*waist))
    Om13 = Omega13 * exp(-r_sq/(2*waist*waist))
    dy[1] = (-Gamma/2)*x[1]-(Gamma/2)*x[2]+(im*conj(Om13)/2)*x[5]-(im*Om13/2)*x[6]+Gamma/2
    dy[2] = (-Gamma/2)*x[1]-(Gamma/2)*x[2]+(im*conj(Om23)/2)*x[7]-(im*Om23/2)*x[8]+Gamma/2
    dy[3] = -gamma21tilde*x[3]+(im*conj(Om23)/2)*x[5]-(im*Om13/2)*x[8]
    dy[4] = -conj(gamma21tilde)*x[4] - (im*Om23/2)*x[6] + (im*conj(Om13)/2)*x[7]
    dy[5] = im*Om13*x[1] + (im*Om13/2)*x[2] + (im*Om23/2)*x[3] - gamma31tilde*x[5]-im*Om13/2
    dy[6] = -im*conj(Om13)*x[1]-im*(conj(Om13)/2)*x[2]-(im*conj(Om23)/2)*x[4]-conj(gamma31tilde)*x[6]+im*conj(Om13)/2
    dy[7] = (im*Om23/2)*x[1]+im*Om23*x[2]+(im*Om13/2)*x[4]-gamma32tilde*x[7] - im*Om23/2
    dy[8] = (-im*conj(Om23)/2)*x[1]-im*conj(Om23)*x[2]-(im*conj(Om13)/2)*x[3]-conj(gamma32tilde)*x[8]+im*conj(Om23)/2
end

paths = [[] for i=1:N_real]
xs = [[] for i=1:N_real]
ts = [[] for i=1:N_real]
v_perps = zeros(Float32, N_real)
function prob_func(prob, i, repeat)
    # change seed of random number generation as the solver uses multithreading
    iinit, jinit, ifinal, jfinal = choose_points()
    vz = draw_vz(v)
    v_perp = sqrt(v^2 - vz^2)
    xinit = jinit*window/N_grid
    yinit = iinit*window/N_grid
    xfinal = jfinal*window/N_grid
    yfinal = ifinal*window/N_grid
    # velocity unit vector
    if v_perp != 0
        u0 = xfinal-xinit
        u1 = yfinal-yinit
        norm = hypot(u0, u1)
        u0 /= norm
        u1 /= norm
        new_tfinal = hypot((xfinal-xinit), (yfinal-yinit))/v_perp
    else
        u0 = u1 = 0
        new_tfinal = hypot((xfinal-xinit), (yfinal-yinit))/abs(vz)
    end
    new_p = ComplexF32[v_perp + 0*im, u0 + 0*im, u1 + 0*im, xinit + 0*im, yinit + 0*im,
             Gamma, Omega13, Omega23, gamma21tilde, gamma31tilde - im*k*vz,
             gamma32tilde - im*k*vz, waist, r0]
    new_tspan = (0.0f0, Float32(new_tfinal))
    tsave = collect(LinRange{Float32}(0.0f0, new_tfinal, 1000))
    remake(prob, p=new_p, tspan=new_tspan, saveat=tsave)
end
# instantiate a problem
p = ComplexF32[1.0 + 0*im, 1.0 + 0*im, 1.0 + 0*im, 1.0 + 0*im, 1.0 + 0*im,
     Gamma, Omega13,
     Omega23, gamma21tilde,
     gamma31tilde - im*k*0.0,
     gamma32tilde - im*k*0.0, waist, r0]
tspan = (0.0f0, 1.0f0)
tsave = collect(LinRange{Float32}(0.0f0, 1.0f0, 1000))
prob = ODEProblem{true}(f!, x0, tspan, p, saveat=tsave)
ensembleprob = EnsembleProblem(prob, prob_func=prob_func)
@time sol = solve(ensembleprob, BS3(), EnsembleGPUArray(), trajectories=N_real)
```

--- a/src/DiffEqGPU.jl
+++ b/src/DiffEqGPU.jl
@@ -313,7 +313,7 @@ This introduces the following limitations on its usage:
   multi-GPU tutorial for more details.
 
 !!! warn
-    Callbacks with `terminate!` does not work well with `EnsembleGPUArray` as the entire 
+    Callbacks with `terminate!` does not work well with `EnsembleGPUArray` as the entire
     integration will hault when any of the trajectories hault. Use with caution.
 
 ## Example
@@ -430,6 +430,11 @@ monteprob = EnsembleProblem(prob, prob_func = prob_func, safetycopy = false)
 struct EnsembleGPUKernel <: EnsembleKernelAlgorithm
     cpu_offload::Float64
 end
+
+cpu_alg = Dict(GPUTsit5 => (GPUSimpleTsit5(), GPUSimpleATsit5()),
+               GPUVern7 => (GPUSimpleVern7(), GPUSimpleAVern7()),
+               GPUVern9 => (GPUSimpleVern9(), GPUSimpleAVern9()))
+
 # Work around the fact that Zygote cannot handle the task system
 # Work around the fact that Zygote isderiving fails with constants?
 function EnsembleGPUArray()
@@ -476,9 +481,9 @@ function SciMLBase.__solve(ensembleprob::SciMLBase.AbstractEnsembleProblem,
         cpu_II = (gpu_trajectories + 1):trajectories
         _alg = if typeof(alg) <: Union{GPUTsit5, GPUVern7, GPUVern9}
             if adaptive == false
-                GPUSimpleTsit5()
+                cpu_alg[typeof(alg)][1]
             else
-                GPUSimpleATsit5()
+                cpu_alg[typeof(alg)][2]
             end
         elseif typeof(alg) <: Union{GPUEM}
             if adaptive == false
@@ -585,15 +590,24 @@ function batch_solve(ensembleprob, alg,
             ensembleprob.prob_func(ensembleprob.prob, i, 1)
         end
     end
-    @assert all(Base.Fix2((prob1, prob2) -> isequal(prob1.tspan, prob2.tspan), probs[1]),
-                probs)
     @assert !isempty(I)
     #@assert all(p->p.f === probs[1].f,probs)
 
     if ensemblealg isa EnsembleGPUKernel
+        # Using inner saveat requires all of them to be of same size,
+        # because the dimension of CuMatrix is decided by it.
+        # The columns of it are accessed at each thread.
+        @assert all(Base.Fix2((prob1, prob2) -> isequal(sizeof(get(prob1.kwargs, :saveat,
+                                                                   nothing)),
+                                                        sizeof(get(prob2.kwargs, :saveat,
+                                                                   nothing))), probs[1]),
+                    probs)
         if typeof(alg) <: Union{GPUTsit5, GPUVern7, GPUVern9, GPUEM}
+            # Get inner saveat if global one isn't specified
+            _saveat = get(probs[1].kwargs, :saveat, nothing)
+            saveat = _saveat === nothing ? get(kwargs, :saveat, nothing) : _saveat
             solts, solus = batch_solve_up_kernel(ensembleprob, probs, alg, ensemblealg, I,
-                                                 adaptive; kwargs...)
+                                                 adaptive; saveat = saveat, kwargs...)
 
             sols = ensembleprob.prob isa SDEProblem ?
                    Array{RODESolution}(undef, length(probs)) :
@@ -621,6 +635,9 @@ function batch_solve(ensembleprob, alg,
             error("We don't have solvers implemented for this algorithm yet")
         end
     else
+        @assert all(Base.Fix2((prob1, prob2) -> isequal(prob1.tspan, prob2.tspan),
+                              probs[1]),
+                    probs)
         u0 = reduce(hcat, Array(probs[i].u0) for i in 1:length(I))
         p = reduce(hcat,
                    probs[i].p isa SciMLBase.NullParameters ? probs[i].p : Array(probs[i].p)

--- a/src/DiffEqGPU.jl
+++ b/src/DiffEqGPU.jl
@@ -597,11 +597,25 @@ function batch_solve(ensembleprob, alg,
         # Using inner saveat requires all of them to be of same size,
         # because the dimension of CuMatrix is decided by it.
         # The columns of it are accessed at each thread.
-        @assert all(Base.Fix2((prob1, prob2) -> isequal(sizeof(get(prob1.kwargs, :saveat,
+        if !all(Base.Fix2((prob1, prob2) -> isequal(prob1.tspan, prob2.tspan),
+                          probs[1]),
+                probs)
+            if !iszero(ensemblealg.cpu_offload)
+                error("Different time spans in an Ensemble Simulation with CPU offloading is not supported yet.")
+            end
+            if get(probs[1].kwargs, :saveat, nothing) === nothing && !adaptive &&
+               get(kwargs, :save_everystep, true)
+                error("Using different time-spans require either turning off save_everystep or using saveat. If using saveat, it should be of same length across the ensemble.")
+            end
+            if !all(Base.Fix2((prob1, prob2) -> isequal(sizeof(get(prob1.kwargs, :saveat,
                                                                    nothing)),
                                                         sizeof(get(prob2.kwargs, :saveat,
                                                                    nothing))), probs[1]),
                     probs)
+                error("Using different saveat in EnsembleGPUKernel requires all of them to be of same length. Use saveats of same size only.")
+            end
+        end
+
         if typeof(alg) <: Union{GPUTsit5, GPUVern7, GPUVern9, GPUEM}
             # Get inner saveat if global one isn't specified
             _saveat = get(probs[1].kwargs, :saveat, nothing)

--- a/src/perform_step/gpu_em_perform_step.jl
+++ b/src/perform_step/gpu_em_perform_step.jl
@@ -10,6 +10,10 @@ function em_kernel(probs, _us, _ts, dt,
     ts = @inbounds view(_ts, :, i)
     us = @inbounds view(_us, :, i)
 
+    _saveat = get(prob.kwargs, :saveat, nothing)
+
+    saveat = _saveat === nothing ? saveat : _saveat
+
     f = prob.f
     g = prob.g
     u0 = prob.u0

--- a/src/perform_step/gpu_tsit5_perform_step.jl
+++ b/src/perform_step/gpu_tsit5_perform_step.jl
@@ -78,6 +78,10 @@ function tsit5_kernel(probs, _us, _ts, dt, callback, tstops, nsteps,
     ts = @inbounds view(_ts, :, i)
     us = @inbounds view(_us, :, i)
 
+    _saveat = get(prob.kwargs, :saveat, nothing)
+
+    saveat = _saveat === nothing ? saveat : _saveat
+
     integ = gputsit5_init(prob.f, false, prob.u0, prob.tspan[1], dt, prob.p, tstops,
                           callback, save_everystep, saveat)
 
@@ -232,6 +236,10 @@ function atsit5_kernel(probs, _us, _ts, dt, callback, tstops, abstol, reltol,
     ts = @inbounds view(_ts, :, i)
     us = @inbounds view(_us, :, i)
     # TODO: optimize contiguous view to return a CuDeviceArray
+
+    _saveat = get(prob.kwargs, :saveat, nothing)
+
+    saveat = _saveat === nothing ? saveat : _saveat
 
     u0 = prob.u0
     tspan = prob.tspan

--- a/src/perform_step/gpu_vern7_perform_step.jl
+++ b/src/perform_step/gpu_vern7_perform_step.jl
@@ -86,6 +86,10 @@ function vern7_kernel(probs, _us, _ts, dt, callback, tstops, nsteps,
     ts = @inbounds view(_ts, :, i)
     us = @inbounds view(_us, :, i)
 
+    _saveat = get(prob.kwargs, :saveat, nothing)
+
+    saveat = _saveat === nothing ? saveat : _saveat
+
     integ = gpuvern7_init(prob.f, false, prob.u0, prob.tspan[1], dt, prob.p, tstops,
                           callback, save_everystep, saveat)
 
@@ -257,6 +261,10 @@ function avern7_kernel(probs, _us, _ts, dt, callback, tstops, abstol, reltol,
     ts = @inbounds view(_ts, :, i)
     us = @inbounds view(_us, :, i)
     # TODO: optimize contiguous view to return a CuDeviceArray
+
+    _saveat = get(prob.kwargs, :saveat, nothing)
+
+    saveat = _saveat === nothing ? saveat : _saveat
 
     u0 = prob.u0
     tspan = prob.tspan

--- a/src/perform_step/gpu_vern9_perform_step.jl
+++ b/src/perform_step/gpu_vern9_perform_step.jl
@@ -113,6 +113,10 @@ function vern9_kernel(probs, _us, _ts, dt, callback, tstops, nsteps,
     ts = @inbounds view(_ts, :, i)
     us = @inbounds view(_us, :, i)
 
+    _saveat = get(prob.kwargs, :saveat, nothing)
+
+    saveat = _saveat === nothing ? saveat : _saveat
+
     integ = gpuvern9_init(prob.f, false, prob.u0, prob.tspan[1], dt, prob.p, tstops,
                           callback, save_everystep, saveat)
 
@@ -316,6 +320,10 @@ function avern9_kernel(probs, _us, _ts, dt, callback, tstops, abstol, reltol,
     ts = @inbounds view(_ts, :, i)
     us = @inbounds view(_us, :, i)
     # TODO: optimize contiguous view to return a CuDeviceArray
+
+    _saveat = get(prob.kwargs, :saveat, nothing)
+
+    saveat = _saveat === nothing ? saveat : _saveat
 
     u0 = prob.u0
     tspan = prob.tspan

